src/main/scala/analysis/SSAForm.scala

```diff
@@ -0,0 +1,99 @@
+        blockBasedMappings.getOrElseUpdate((block, v.name),
+        context.getOrElseUpdate((proc, v.name), mutable.Set(getMax(v.name)))
+      ))
+    })
+  }
+
+  def applySSA(program: Program): Unit = {
+    for (proc <- program.procedures) {
+      val visitedBlocks = mutable.Set[Block]()
+      val stack = mutable.Stack[Block]()
+
+      // Start with the entry block
+      if (proc.entryBlock.isDefined) {
+        stack.push(proc.entryBlock.get)
+      }
+
+      while (stack.nonEmpty) {
+        val currentBlock = stack.pop()
+
+        if (!visitedBlocks.contains(currentBlock)) {
+          visitedBlocks.add(currentBlock)
+
+          for (stmt <- currentBlock.statements) {
+            println(stmt)
+            stmt match {
+              case localAssign: LocalAssign => {
+                transformVariables(localAssign.rhs.variables, currentBlock, proc)
+                val maxVal = varMaxTracker.getOrElseUpdate(localAssign.lhs.name, 0)
+                blockBasedMappings((currentBlock, localAssign.lhs.name)) = mutable.Set(maxVal)
+
+                localAssign.lhs.ssa_id.clear()
+                localAssign.lhs.ssa_id.addAll(blockBasedMappings((currentBlock, localAssign.lhs.name)))
+
+                varMaxTracker(localAssign.lhs.name) = blockBasedMappings((currentBlock, localAssign.lhs.name)).max + 1
+              }
+              case memoryAssign: MemoryAssign => {
+                transformVariables(memoryAssign.lhs.variables, currentBlock, proc)
+                transformVariables(memoryAssign.rhs.variables, currentBlock, proc)
+              }
+              case assume: Assume => {
+                transformVariables(assume.body.variables, currentBlock, proc)
+              } // no required for analyses
+              case assert: Assert => {
+                transformVariables(assert.body.variables, currentBlock, proc)
+              } // no required for analyses
+              case _ => throw new RuntimeException("No SSA form for " + stmt.getClass + " yet")
+            }
+          }
+          currentBlock.jump match {
+            case directCall: DirectCall =>
+              // TODO: transfers the whole context but it could be using ANR and RNA to transfer only the relevant context
+              varMaxTracker.keys.foreach(varr => {
+                //context((directCall.target, varr)) = context((directCall.target, varr)) ++ blockBasedMappings(block, varr)
+                context.getOrElseUpdate((directCall.target, varr), mutable.Set()) ++= blockBasedMappings((currentBlock, varr))
+              })
+            case indirectCall: IndirectCall =>
+              transformVariables(indirectCall.target.variables, currentBlock, proc)
+            case goTo: GoTo =>
+              goTo.targets.foreach(b => {
+                varMaxTracker.keys.foreach(varr => {
+                  blockBasedMappings((b, varr)) = blockBasedMappings(b, varr) ++ blockBasedMappings(currentBlock, varr)
+                })
+              })
```

Comment:
This is dependent on visiting blocks after all of their predecessors, which no attempt is made to guarantee. If a block is visited before all of its predecessors, then its SSA values will be incorrect.

For example, if we have the following blocks:

A goto B, C
C goto B

There are two possible orders the blocks could be visited in (as the next blocks are stored in a set) - either A, B, C or A, C, B. If they are visited in the order A, B, C, then the visit to B will incorrectly miss using C's SSA values.